hivemq-edge/src/main/java/com/hivemq/protocols/ProtocolAdapterWrapper.java

```diff
@@ -88,40 +94,110 @@ public ProtocolAdapterWrapper(
-            return startWriting(writingEnabled, protocolAdapterWritingService);
-        }).thenApply(r -> {
-            setRuntimeStatus(ProtocolAdapterState.RuntimeStatus.STARTED);
-            return null;
-        });
-
+    public @NotNull CompletableFuture<Boolean> startAsync(
+            final boolean writingEnabled,
+            final @NotNull ModuleServices moduleServices) {
+        final var currentFuture = startFutureRef.get();
+        if(currentFuture == null) {
+            synchronized (startFutureRef) {
+                final var currentFutureDoubleCheck = startFutureRef.get();
+                if(currentFutureDoubleCheck != null) {
+                    return currentFutureDoubleCheck;
+                }
+                initStartAttempt();
+                final ProtocolAdapterStartOutputImpl output = new ProtocolAdapterStartOutputImpl();
+                final ProtocolAdapterStartInputImpl input = new ProtocolAdapterStartInputImpl(moduleServices);
+
+                final CompletableFuture<Boolean> startFuture = new CompletableFuture<>();
+                startFutureRef.set(startFuture);
+
+                CompletableFuture
+                        .supplyAsync(() -> {
+                            adapter.start(input, output);
+                            return output.getStartFuture();
+                        })
+                        .thenCompose(Function.identity())
+                        .handle((ignored, t) -> {
+                            if(t == null) {
+                                createAndSubscribeTagConsumer();
+                                startPolling(protocolAdapterPollingService, input.moduleServices().eventService());
+                                return startWriting(writingEnabled, protocolAdapterWritingService)
+                                        .thenApply(v -> {
+                                            log.info("Successfully started adapter with id {}", adapter.getId());
+                                            setRuntimeStatus(ProtocolAdapterState.RuntimeStatus.STARTED);
+                                            startFutureRef.get().complete(true);
+                                            startFutureRef.set(null);
+                                            return true;
+                                        });
+                            } else {
+                                log.error("Error starting protocol adapter", t);
+                                stopPolling(protocolAdapterPollingService);
+                                return stopWriting(protocolAdapterWritingService)
+                                        .thenApply(v -> {
+                                            log.error("Error starting adapter with id {}", adapter.getId(), t);
+                                            setRuntimeStatus(ProtocolAdapterState.RuntimeStatus.STOPPED);
+                                            startFutureRef.get().complete(false);
```

Comment:
I wonder this is really safe because it is out of `synchronized (startFutureRef)`. If `startAsync()` was called twice at the same time, what would happen?

Reply:
That's why I use doublecheck logging.
The double check (check atomic -> sync on atomic -> check atomic again) prevents a second future from being created for the stop and the start case.
I first check the content of startFutureRef, then do the sync, then check again.
Only then the running future is created.

Since the assign of the AtomicReference happens in the sync block it is ensured no other Completablefuture will ever be created for starting or stopping.

Do see any way this assertion could be broken?